tiledb/api/src/context.rs

```diff
@@ -68,95 +68,110 @@ impl Context {
         let res = unsafe { ffi::tiledb_ctx_alloc(cfg.capi(), &mut c_ctx) };
         if res == ffi::TILEDB_OK {
             Ok(Context {
-                raw: RawContext::Owned(c_ctx),
+                raw: Arc::new(Mutex::new(RawContext { raw: c_ctx })),
             })
         } else {
             Err(Error::LibTileDB(String::from("Could not create context")))
         }
     }
 
-    pub fn get_stats(&self) -> TileDBResult<String> {
-        let mut c_json = std::ptr::null_mut::<std::os::raw::c_char>();
-        let res = unsafe {
-            ffi::tiledb_ctx_get_stats(
-                *self.raw,
-                &mut c_json as *mut *mut std::os::raw::c_char,
-            )
+    pub fn capi_call<Callable>(&self, action: Callable) -> TileDBResult<()>
+    where
+        Callable: FnOnce(*mut ffi::tiledb_ctx_t) -> i32,
+    {
+        // Scope the lock so we can get the error if one happens.
+        let res = {
+            // Docs say that the error result from a poisoned lock are usually
+            // just propagated to all threads by just calling unwrap on locks.
+            let raw_ctx = self.raw.lock().unwrap();
+            action(raw_ctx.raw)
         };
+
         if res == ffi::TILEDB_OK {
-            assert!(!c_json.is_null());
-            let raw = RawStatsString::Owned(c_json);
-            let json = unsafe { std::ffi::CStr::from_ptr(*raw) };
-            Ok(String::from(json.to_string_lossy()))
+            Ok(())
         } else {
             Err(self.expect_last_error())
         }
     }
 
+    pub fn get_stats(&self) -> TileDBResult<String> {
+        let mut c_json: *mut std::ffi::c_char = out_ptr!();
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_get_stats(
+                ctx,
+                &mut c_json as *mut *mut std::ffi::c_char,
+            )
+        })?;
+
+        assert!(!c_json.is_null());
+        let raw = RawStatsString::Owned(c_json);
+        let json = unsafe { std::ffi::CStr::from_ptr(*raw) };
+        Ok(String::from(json.to_string_lossy()))
+    }
+
     pub fn get_config(&self) -> TileDBResult<Config> {
         let mut c_cfg: *mut ffi::tiledb_config_t = out_ptr!();
-        let res = unsafe { ffi::tiledb_ctx_get_config(*self.raw, &mut c_cfg) };
-        if res == ffi::TILEDB_OK {
-            Ok(Config {
-                raw: RawConfig::Owned(c_cfg),
-            })
-        } else {
-            Err(self.expect_last_error())
-        }
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_get_config(ctx, &mut c_cfg)
+        })?;
+
+        Ok(Config {
+            raw: RawConfig::Owned(c_cfg),
+        })
     }
 
-    pub fn get_last_error(&self) -> Option<Error> {
+    // This is a bit awkard. We *expect* this to return an Err variant of the
```

Comment:
Are your concerns allayed if these are changed to `pub(crate)`?

All the API calls should check these, there should be no reason for a user to have to do so.

Reply:
I was mostly just not liking my weird use of TileDBResult here. It was just "Wait, should an error be returned as an `Ok` or an `Err`?" sort of weirdness. As you pointed out we'd want to make these called while the lock is held if we go that direction.

Reply:
I did this differently and just swallowed the error like we were before so its same as it used to be.